File: code/py/ch10/10.03.py
### 10.3.2 Pythonの場合

import pandas as pd
my_data = pd.DataFrame({
  'answer': [  0,   1,   1,   0,   1,   0,    1,   0,   0,   1],  # 正解
  'prob':   [0.7, 0.8, 0.3, 0.4, 0.9, 0.6, 0.99, 0.1, 0.2, 0.5]}) # 陽性確率

from sklearn.metrics import roc_curve
fpr, tpr, _ = roc_curve(my_data.answer, my_data.prob)

import matplotlib.pyplot as plt
plt.plot(fpr, tpr)
plt.xlabel('False positive rate')
plt.ylabel('True positive rate')

from sklearn.metrics import auc
auc(fpr, tpr)
#> 0.8

File: code/R/ch10/10.03.R
### 10.3.1 質的入力変数を含むデータでの学習

library(caret)
library(PRROC)
library(tidyverse)

my_url <- str_c("https://raw.githubusercontent.com",
                "/taroyabuki/fromzero/master/data/titanic.csv")
my_data <- read_csv(my_url)

head(my_data)
#> # A tibble: 6 x 4
#>   Class Sex   Age   Survived
#>   <chr> <chr> <chr> <chr>   
#> 1 1st   Male  Child Yes     
#> 2 1st   Male  Child Yes     
#> 3 1st   Male  Child Yes     
#> 4 1st   Male  Child Yes     
#> 5 1st   Male  Child Yes     
#> 6 1st   Male  Adult No

### 10.3.2 分類木

my_model <- train(form = Survived ~ ., data = my_data, method = "rpart2",
                  tuneGrid = data.frame(maxdepth = 2),
                  trControl = trainControl(method = "LOOCV"))

#### 10.3.2.1 分類木の描画

rpart.plot::rpart.plot(my_model$finalModel, extra = 1)

#### 10.3.2.2 分類木の評価

my_model$results
#>   maxdepth  Accuracy     Kappa
#> 1        2 0.7832803 0.4096365

y <- my_data$Survived
tmp <- my_model %>% predict(newdata = my_data, type = "prob")
y_score <- tmp$Yes

my_roc <- roc.curve(scores.class0 = y_score[y == "Yes"],
                    scores.class1 = y_score[y == "No"],
                    curve = TRUE)
my_roc$auc
#> [1] 0.7114887

my_roc %>% plot(xlab = "False Positive Rate",
                ylab = "True Positive Rate",
                legend = FALSE)

